client-api: Support appservice login type on /register

johannescpk opened the pull request:
Relevant part of the spec: https://matrix.org/docs/spec/application_service/r0.1.2#server-admin-style-permissions

Blocks https://github.com/matrix-org/matrix-rust-sdk/pull/204

johannescpk commented on ruma-client-api/src/r0/account/register.rs:
```diff
@@ -13,7 +13,7 @@ ruma_api! {
         path: "/_matrix/client/r0/register",
         rate_limited: true,
-        authentication: None,
+        authentication: AccessToken,
```
Since registering with [admin style permissions](https://matrix.org/docs/spec/application_service/r0.1.2#server-admin-style-permissions) requires providing the access_token in the header additionally to the login type I'm not actually sure how to properly do that here. Maybe it needs another Request-variant?

jplatte commented:
This would break regular usage of this endpoint. I have an idea of how to handle this and the optional auth on the profile info endpoints though: In `ruma-api/src/lib.rs`, replace `access_token: Option<&str>` with something like

```rust
pub enum SendAccessToken<'a> {
    IfRequired(&'a str),
    Always(&'a str),
    No,
}
```

and update the macro expansion of `ruma_api!` accordingly: `IfRequired` and `No` represent the existing `Some` and `None` cases, `Always` adds the access token regardless of whether the endpoint requires authentication or not.

jplatte commented:
Actually I guess you can just reuse this branch / PR for that change, no need to close.

johannescpk committed:
client-api: Support appservice login type on /register

johannescpk commented on ruma-api-macros/src/api/request/outgoing.rs:
```diff
@@ -131,23 +131,31 @@ impl Request {
-                header_kvs.extend(quote! {
+            let attrs = &auth.attrs;
+            let auth_value = &auth.value.to_string();
+            header_kvs.extend(quote! {{
```
Using `{{` here to have shadowing of `access_token` in it's own scope as it makes problems otherwise (and might lead to unexpected behavior if extended in the future)

jplatte commented:
Can you rebase to pick up the recent CI fixes?

johannescpk commented:
Done

jplatte commented on ruma-api-macros/src/api/request/outgoing.rs:
```diff
@@ -131,23 +131,31 @@ impl Request {
+            header_kvs.extend(quote! {{
+                let access_token = match access_token {
+                    #ruma_api::SendAccessToken::IfRequired(access_token) => {
+                        if #auth_value == "AccessToken" {
```
This is extra code for the compiler to constant-fold for every single invocation of the macro. This check should be outside of `quote!`.

johannescpk replied:
So like moving just `#auth_value == "AccessToken"` outside of `quote!` and doing `if boolean_result {` here? Because moving out the whole match doesn't work since it needs the `access_token`

johannescpk replied:
The former makes sense anyway, so done that

jplatte commented on ruma-api-macros/src/api/request/outgoing.rs:
```diff
@@ -131,23 +131,31 @@ impl Request {
+                        } else {
+                            None
+                        }
+                    },
```
That is not what I meant. You're still emitting an `if else` expression even though the path that will be taken is know already at the point of the code being generated. I was thinking of something like

```suggestion
            let if_required_expr = if auth.value == "AccessToken" {
                quote! { Some(access_token) }
            } else {
                quote! { None }
            };
            header_kvs.extend(quote! {{
                let access_token = match access_token {
                    #ruma_api::SendAccessToken::IfRequired(access_token) => #if_required_expr,
```

johannescpk replied:
The `access_token` is not available outside of that `quote!`

johannescpk replied:
Ahhh, now I see what you mean, sorry

jplatte replied:
It doesn't need to be.

jplatte replied:
Identifiers in `quote!` are not hygienic in any way.

johannescpk replied:
Yeah, hadn't noticed that you quote!ed the Option as well, which is smart. Should read more carefully. Looks better now!

johannescpk committed:
api: Replace access_token Option with SendAccessToken enum

jplatte commented on ruma-api-macros/src/api/request/outgoing.rs:
```diff
@@ -131,23 +131,29 @@ impl Request {
-                            &::std::format!(
-                                "Bearer {}",
-                                access_token.ok_or(
-                                    #ruma_api::error::IntoHttpError::NeedsAuthentication
```
Looking at the overall logic again, it seems like we have lost the logic that if the user doesn't specify an access token (now `SendAccessToken::No`) but one is required according to the endpoint definition, we return `Err(NeedsAuthentication)` without even sending the request.

johannescpk replied:
My interpretation, and what the docs for `SendAccessToken::No` currently suggest, is that `SendAccessToken::No` is supposed to simply never add an `access_token` even if the endpoint requires it. While `SendAccessToken::IfRequired` adds the given `access_token` only if the endpoint requires it, and none otherwise. So it seems there is no `Err(NeedsAuthentication)` case any longer?

johannescpk replied:
Maybe `SendAccessToken::Never` would be more clear?

jplatte replied:
> My interpretation, and what the docs for `SendAccessToken::No` currently suggest, is that `SendAccessToken::No` is supposed to simply never add an `access_token` even if the endpoint requires it.

I don't think we have a use case for that though. What `access_token: None` currently means is that the client simply doesn't have an access token yet, i.e. the user is not logged in. Maybe we need to have a different enum name than `SendAccessToken` to make this clearer, but that's the only thing I've been able to come up with.

johannescpk replied:
So `SendAccessToken::None` with documentation and code saying, that if the endpoint requires the token, it'll return an error?

jplatte replied:
If you think that's better than `SendAccessToken::No`, sure.

johannescpk replied:
Updated. Yeah, I feel like `None` better suggests that an access_token simply isn't available (in combination with the enum name), while `No` sounds more like, no, you never going to send that access token :D

jplatte reviewed (commented):
If you want, you can add a second commit to the PR that removes `unstable-synapse-quirks` again, since the only thing it was doing was declaring an endpoint as requiring authentication that doesn't need it per spec, which now is not required anymore since the access token can be force-added for that endpoint with `SendAccessToken::Always`.

johannescpk committed:
client-api: Remove unstable-synapse-quirks feature

Only worked around `authentication` sometimes being required,
which now can be achieved with `SendAccessToken`

jplatte commented on ruma-api-macros/src/api/request/outgoing.rs:
```diff
@@ -131,23 +131,36 @@ impl Request {
+                let access_token = match access_token {
+                    #ruma_api::SendAccessToken::IfRequired(access_token) => #if_required_expr,
+                    #ruma_api::SendAccessToken::Always(access_token) => Some(access_token),
+                    #ruma_api::SendAccessToken::None => if #auth_required {
```
Now you again introduced an if-else with a constant condition 😅

johannescpk replied:
Right :sweat_smile:

johannescpk replied:
Gave it another try :sweat_smile:

jplatte commented on ruma-api-macros/src/api/request/outgoing.rs:
```diff
@@ -131,23 +131,38 @@ impl Request {
+                quote! { return Err(#ruma_api::error::IntoHttpError::NeedsAuthentication) }
+            } else {
+                quote! { None }
+            };
```
Wdyt about this?

```suggestion
            let (if_required_expr, non-expr) = if auth.value == "AccessToken" {
                (
                    quote! { Some(access_token) },
                    quote! { return Err(#ruma_api::error::IntoHttpError::NeedsAuthentication) }
                )
            } else {
                (quote! { None }, quote! { None })
            };
```
